### PR TITLE
Implement `value` property on `<select>`.

### PR DESCRIPTION
**Pull request by frewsxcv**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13783)

### CODE REVIEW

File: components/script/dom/htmlselectelement.rs [hunk 1]
```diff
@@ -258,6 +258,36 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
     fn Remove(&self) {
         self.upcast::<Element>().Remove()
     }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-value
+    fn Value(&self) -> DOMString {
+        self.upcast::<Node>()
+            .traverse_preorder()
```
Comment (nox) [inline components/script/dom/htmlselectelement.rs]: This doesn't seem correct to me. What about this tree?

```html
<select>
  <div>
    <option value="I should be ignored" selected="selected">
  </div>
</select>
```

File: components/script/dom/htmlselectelement.rs [hunk 2]
```diff
@@ -258,6 +258,36 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
+            .next()
+            .unwrap_or_default()
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-value
+    fn SetValue(&self, value: DOMString) {
+        let mut opt_iter = self.upcast::<Node>()
+                               .traverse_preorder()
+                               .filter_map(Root::downcast::<HTMLOptionElement>);
+        while let Some(opt) = opt_iter.next() {
```
Comment (nox) [inline components/script/dom/htmlselectelement.rs]: Why a `while` loop?

File: components/script/dom/htmlselectelement.rs [hunk 3]
```diff
@@ -258,6 +258,36 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
+            .filter(|opt_elem| opt_elem.Selected())
+            .map(|opt_elem| opt_elem.Value())
+            .next()
+            .unwrap_or_default()
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-value
+    fn SetValue(&self, value: DOMString) {
+        let mut opt_iter = self.upcast::<Node>()
+                               .traverse_preorder()
```
Comment (nox) [inline components/script/dom/htmlselectelement.rs]: Same as above.

File: components/script/dom/htmlselectelement.rs [hunk 4]
```diff
@@ -258,6 +258,36 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
+        while let Some(opt) = opt_iter.next() {
+            if opt.Value() == value {
+                opt.set_selectedness(true);
+                opt.set_dirtiness(true);
+                break;
+            }
+            opt.set_selectedness(false);
+        }
+        // Reset remaining <option> elements
+        for ref opt in opt_iter {
```
Comment (nox) [inline components/script/dom/htmlselectelement.rs]: Why is the `ref` necessary here?

**Review by nox — CHANGES_REQUESTED** [review]
Some changes are required. We should also add a test to check that we correctly skip options that are not children of `select` or `optgroup`.

File: components/script/dom/htmlselectelement.rs (hunk 1 above)
Comment (frewsxcv) [inline components/script/dom/htmlselectelement.rs]: Same algorithm is done elsewhere in the file and was going to take care of it with https://github.com/servo/servo/issues/13763

File: components/script/dom/htmlselectelement.rs (hunk 2 above)
Comment (frewsxcv) [inline components/script/dom/htmlselectelement.rs]: `opt_iter` gets moved when I tried a `for opt in opt_iter`

File: components/script/dom/htmlselectelement.rs (hunk 4 above)
Comment (frewsxcv) [inline components/script/dom/htmlselectelement.rs]: `ref` removed in the latest force push.

File: components/script/dom/htmlselectelement.rs (hunk 3 above)
Comment (frewsxcv) [inline components/script/dom/htmlselectelement.rs]: https://github.com/servo/servo/pull/13783#discussion_r83550821

File: components/script/dom/htmlselectelement.rs (hunk 2 above)
Comment (frewsxcv) [inline components/script/dom/htmlselectelement.rs]: Just kidding, forgot `Iterator::by_ref` exists.

**Comment by frewsxcv** [comment]
@nox Comments have been addressed. Also did a rewriting of the `SetValue` flow.

**Pull request merged**